# <center> QCM of CoDaDri - Academic Year 2021/2022 - 12th November 2021</center>

## Rules of the game

Before starting, here it is a short  summary of things that you should know. In the following, there are 19 questions. For each question you have four choices: three are wrong and one is correct. If you check the correct one you get a point, but if you are wrong you lose half a point. No answer zero points. Good Luck!


- **Trial Question:**

Consider the following programs:


In [ ]:
#Program A:

for i in range(10):
    print (i)

In [ ]:
#Program B:

for i in range(10):
print (i)

Which of the above programs (A, B) prints all integers from 
0 to 9 (including these numbers)?

[a] Program A does the job.

[b] Program B does the job.

[c] Both Program A and Program B do the job.

[d] All programs are  wrong.

If you check [a] - correct answer. [+ 1 point]
If you check [b] - wrong answer (indentation missing). [-0.25 points]
If you check [c] - wrong answer. [-0.25 points]
if you check [d] - wrong answer. [-0.25 points]

---

**Question 1**

Your goal is to sample evenly the population of 9 possible positions. Consider a Markov chain strategy with a Random Surfer moving at each time-step from its position to a new one, uniformly selected from a list of neighbors. Consider the following  neighbor list: 

In [4]:
import random
neighbor = {1 : [2, 4, 2, 4], 2 : [1, 3, 5, 1], 3 : [2, 6, 2, 1],
            4 : [5, 1, 5, 1], 5 : [6, 8, 4, 2], 6 : [9, 9, 5, 3],
            7 : [8, 4, 9, 4], 8 : [9, 7, 7, 5], 9 : [6, 8, 8, 6]}
all_pos = []
N_trials = 10000
pos = 9
for iter in range(N_trials):
    pos = random.choice(neighbor[pos])
    all_pos.append(pos)
print(np.unique(all_pos, return_counts=True))   

(array([1, 2, 3, 4, 5, 6, 7, 8, 9]), array([1654, 1561,  642, 1426, 1539,  953,  490,  920,  815]))


[a] The Markov Chain reaches the correct stationary state and the detailed balance is satisfied.

[b] The Markov Chain reaches the correct stationary state, but only the global balance is satified.

[c] The Markov Chain does not satisfy the global balance.

[d] The stationary state is correct, but it is not reached because the Markov Chain is periodic.

**Question 2**

Your goal is to sample evenly the population of 9 possible positions. Consider a Markov chain strategy with a Random Surfer moving at each time-step from its position to a new one, uniformly selected from a list of neighbors. Consider the two following neighbor lists: 

In [ ]:
import random
neighbor_1 =  [1 : [2], 2 : [3], 3 : [4], 4 : [5], 5 : [6], 6 : [7], 7 : [8], 8 : [9], 9 : [1]]

neighbor_2 =  [ 1 : [2, 2, 1], 2 : [3, 3, 2], 3 : [4, 4, 3], 4 : [5, 5, 4], 5 : [6, 6, 5],
               6 : [7, 7, 6], 7 : [8, 8, 7], 8 : [9, 9, 8], 9: [1, 1, 9]]
all_pos = []
N_trials = 10
pos = 9
for iter in range(N_trials):
    pos = random.choice(neighbor_1[pos])
    all_pos.append(pos)
print(all_pos)    

[a] neighbor_1 and neighbor_2 satisfy the detailed balance condition. 

[b] neighbor_1 and neighbor_2 satisfy the global balance condition.

[c] Using neighbor_2 the stationary state is correct, but it is not reached because the Markov Chain is periodic.

[d] Only neighbor_1 is correct as it satisfies the global balance condition and it is aperiodic.

**Question 3**

Your goal is to sample evenly the population of 9 possible positions. Consider a Markov chain strategy with a Random Surfer moving at each time-step from its position to a new one, uniformely selected from a list of neighbors. Consider the following program:

In [5]:
import random
neighbor = {1 : [2, 4, 1, 1], 2 : [3, 5, 1, 2], 3 : [3, 6, 2, 3],
            4 : [5, 7, 4, 1], 5 : [6, 8, 4, 2], 6 : [6, 9, 5, 3],
            7 : [8, 7, 7, 4], 8 : [9, 8, 7, 5], 9 : [9, 9, 8, 6]}
all_pos = []
N_trials = 10
pos = 9
for iter in range(N_trials):
    Epsilon = random.uniform(0.0, 1.0)
    if  Epsilon < 0.25 :
        pos = neighbor[pos][0]
    elif(Epsilon < 0.5) :
        pos = neighbor[pos][1]
    elif(Epsilon < 0.75) :
        pos = neighbor[pos][2]
    else:
        pos = neighbor[pos][3]
    all_pos.append(pos) 

[a] The Markov Chain reaches the correct stationary state and the detailed balance is satisfied.

[b] The Markov Chain reaches the correct stationary state, but only the global balance is satified.

[c] The Markov Chain does not satisfy the global balance.

[d] The stationary state is correct, but it is not reached because the Markov Chain is periodic.

**Question 4**

Your goal is to sample a **weighted** population of 9 possible positions. Let's test your understanding of Metropolis! Consider the following programs 

In [ ]:
import random, math
neighbor = {1 : [2, 4, 1, 1], 2 : [3, 5, 1, 2], 3 : [3, 6, 2, 3],
            4 : [5, 7, 4, 1], 5 : [6, 8, 4, 2], 6 : [6, 9, 5, 3],
            7 : [8, 7, 7, 4], 8 : [9, 8, 7, 5], 9 : [9, 9, 8, 6]}
weight=[1, 3, 2, 1, 2, 2, 3, 1, 1]
all_pos = []
N_trials = 10
pos = 9
#Program 1
for iter in range(N_trials):
    newpos = random.choice(neighbor[pos])
    if random.uniform(0.0, 1.0) < weight(newpos) / weight(pos): 
        pos = new_pos
        all_pos.append(pos)
        
#Program 2
for iter in range(N_trials):
    newpos = random.choice(neighbor[pos])
    if random.uniform(0.0, 1.0) < weight(newpos) / weight(pos): 
        pos = new_pos
    all_pos.append(pos)
    a
#Program 3
for iter in range(N_trials):
    newpos = random.choice(neighbor[pos])
    if random.uniform(0.0, 1.0) < weight(pos) / weight(newpos): 
        pos = new_pos
        all_pos.append(pos) 

[a] Program 1 is a correct implementation of Metropolis

[b] Program 2 is a correct implementation of Metropolis

[c] Program 3 is a correct implementation of Metropolis

[d] The 3 programs are not a correct 


**Question 5**

Your goal is to decide what to do tonight: go to the party at 358 rue the Pyrenees or study for the master? A judicious resolution is to select the second option three times more often than the first one.

In [ ]:
import random
# Program 1
neighbor=['party', 'M2', 'M2', 'M2']
for t in range(20):
    print(random.choice(neighbor))
    
# Program 2
for t in range(20):
    choice='M2'
    if random.uniform(0.0, 1.0) < 0.25: 
       choice='party'
    print(choice)

#Program 3
for t in range(20):
    choice='M2'
    if random.uniform(0.0, 1.0) < 0.5: 
       choice='M2'
    elif random.uniform(0.0, 1.0) < 0.5: 
       choice='party'
    print(choice)

[a] Only program 1 is correct.

[b] Both program 1 and 2 are correct, but 3 is wrong.

[c] They are all wrong.

[d] They are all correct.

**Question 6**

Your goal is to identify the probability density function (pdf) sampled by the code below.

In [34]:
import random
print ((random.random())**(1./2.))

0.4525260591316806


[a] $\pi(x) = \frac{3}{2} \sqrt{x}$ with $0<x<1$ 

[b] $\pi(x) = 2 x$ with $0<x<1$

[c] $\pi(x) = \frac{2}{x^2}$ with $0<x<1$

[d] $\pi(x) = \frac{1}{2 \sqrt{x}}$ with $0<x<1$

**Question 7**

Your goal is to sample points uniformly distributed on a unit sphere. 

In [ ]:
import random, math
#Program 1:
for t in range(1):
    r = [random.gauss(-1.,1.), random.gauss(-1.,1.), random.gauss(-1.,1.)]
    norm=math.sqrt(r[0]**2+r[1]**2+r[2]**2)
    r=[a/norm for a in r]
    print (r)
    
# Program 2:
for t in range(1):
    r = [random.uniform(-1, 1), random.uniform(-1, 1),random.uniform(-1, 1)]
    norm=math.sqrt(r[0]**2 + r[1]**2 + r[2]**2)
    r=[a/norm for a in r]
    print (r)

# Program 3    
for t in range(1): 
    r = [random.uniform(-1, 1), random.uniform(-1, 1),random.uniform(-1, 1)]
    norm=math.sqrt(r[0]**2 + r[1]**2 + r[2]**2)
    if  norm  < 1: 
        r=[a/norm for a in r]
        print (r)

[a] Both program 1 and 3 are wrong, but 2 is correct.

[b] Both program 1 and 2 are wrong, but 3 is correct.

[c] Both program 2 and 3 are wrong, but 1 is correct.

[d] They are all wrong.

**Question 8**

Your goal is to find the maximal energy for a polymer connecting the position $(0,0)$ to the position $(T,0)$. We consider the model of directed polymer in random media defined in Lecture 5 and write the program you find below.

In [ ]:
import numpy as np
T  =  120
E = [-np.infty for n in range(-T,T-1)]
E[0] = 0
for time in range(1,T) :
    for n in range(time+1) :
        x = time - 2*n
        E[x] = max(E[x-1],E[x+1]) + np.random.randn()
print(E[0])        

[a] The program finds the exact value we are looking for.

[b] The program finds an upper bound of the value we are looking for.

[c] The program finds a lower bound of the value we are looking for.

[d] The program finds a good approximation (but not a bound) of the value we are looking for.

**Question 9**

Your goal is to find the energy with minimal absolute value of the energy for a polymer connecting the position $(0,0)$ to the position $(T,0)$. We consider the model of directed polymer in random media defined in Lecture 5 and write the program you find below.

In [ ]:
import numpy as np
T  =  120
E = [-np.infty for n in range(-T,T-1)]
E[0] = 0
for time in range(1,T) :
    for n in range(time+1) :
        x = time - 2*n
        Eold=E[x-1]
        if  abs(E[x-1])>abs(E[x+1]):
            Eold=E[x+1]
        E[x] = Eold + np.random.randn()
print(abs(E[0]))        

[a] The program finds the exact value we are looking for.

[b] The program finds an upper bound of the value we are looking for.

[c] The program finds a lower bound of the value we are looking for.

[d] The program finds a good approximation (but not a bound) of the value we are looking for.

**Question 10**

Your goal is to sample points uniformly distributed on a unit circle. 

In [ ]:
import random, math
#Program 1:
for t in range(1):
    r = [random.gauss(0.,1.), random.gauss(0.,1.)]
    norm=math.sqrt(r[0]**2+r[1]**2)
    r=[a/norm for a in r]
    print (r)
    
# Program 2:
for t in range(1):
    phi=random.uniform(0, 2*math.pi)
    r = [math.cos(phi), math.sin(phi)]
    print (r)

# Program 3     
for t in range(1):
    phi=random.uniform(0, 2*math.pi)
    r = [math.sin(phi), math.cos(phi)]
    print (r)

[a] Only program 1 is correct.

[b] Both program 2 and 3 are correct, but 1 is wrong.

[c] They are all correct.

[d] They are all wrong

**Question 11**

Your goal is to evaluate the following integral:
$$ I(\alpha)=\int_0^1 dx\, \frac{\exp(x)}{x^{\alpha}} \quad \text{with} \; 0<\alpha<1$$
Consider the program below

In [73]:
import random, math
import numpy as np
def O(x):
    return math.exp(x)/x**alpha
alpha = 1
Ntrials=100000
L = []
for t in range(Ntrials):
    x=random.random()
    L.append(O(x))
print("alpha=",alpha,"<I>=", np.mean(L), ",error =", np.std(L)/math.sqrt(Ntrials))

alpha= 1 <I>= 13.92299480399456 ,error = 2.399783710235637



[a] This is a naive strategy and the calculation never converges.

[b] It converges only for $\alpha<1/2$, for $\alpha>1/2$ the variance is divergent.

[c] It converges only for $\alpha>1/2$, for $\alpha<1/2$ the variance is divergent. 

[d] It converges with a finite variance for all $\alpha \in (0,1)$.


**Question 12 (2 points)** 

Your goal is to evaluate the following integral:
$$ I(\alpha)=\int_0^1 dx\, \frac{\exp(x)}{x^{\alpha}} \quad \text{with} \; 0<\alpha<1$$
Use importance sampling with a bias $q(x)=\frac{1-\alpha}{x^\alpha}$. Complete the program below and compute the integral for $\alpha=0.8$ and $\alpha=0.3$ with Ntrials=10000.

In [90]:
import random, math
import numpy as np
alpha = 0.8
Ntrials=10000
L = []
def O(x):
    return np.exp(x)/(1-alpha)
for t in range(Ntrials):
    z=np.random.random()
    x= (z)**(1/(1-alpha))
    L.append(O(x))
print("alpha=",alpha,"<I>=", np.mean(L), ",error I=", np.std(L)/math.sqrt(Ntrials))

alpha= 0.8 <I>= 6.130120716077328 ,error I= 0.019127658285797454


[a] $I(0.8) =6.1 \pm 0.2$,  $I(0.3) =2.26 \pm 0.01$

[b] $I(0.8) =6.13 \pm 0.02$,  $I(0.3) =2.2 \pm 0.1$

[c] $I(0.8) =6.12 \pm 0.02$,  $I(0.3) =2.26 \pm 0.01$

[d] $I(0.8) =6.1 \pm 0.2$,  $I(0.3) =2.2 \pm 0.1$

**Question 13**

Your goal is to evaluate the following integral:
$$ I=\int_{-\infty}^{-10}     \frac{dx}{\sqrt{2 \pi}}  \, \exp(-x^2/2) $$
Use an exponential tilt to evaluate the integral. Modify the program below and compute the integral with Ntrials=100000.

In [110]:
import random,math,numpy
c = 0
t = -10.5
Ntrials=100000
xsum=[]
for i in range(Ntrials):
    dx = random.gauss(c+t,1.)
    if dx < -10:
        num  = (c-dx)*t + t**2/2
        x= math.exp(num)
    else:
        x=0
    xsum.append(x)
xerror = numpy.std(numpy.array(xsum))
xm = numpy.mean(numpy.array(xsum))
print ('simulation', xm,'+/-', xerror/float(Ntrials)**0.5)

simulation 7.613261774058406e-24 +/- 8.715889618753297e-26


[a] $I =7.6  10^{-24 }\pm  10^{-25}$

[b] $I =7.5  10^{-24 }\pm   5 \times 10^{-25}$

[c] $I =7.62  10^{-24 }\pm  10^{-26}$

[d] $I \approx 0.0  \pm  0.0$

**Question 14**

Imagine we found a way to implement a *faster than the clock* algorithm for the Ising model. As a result, we generate a Markov chain with $M$ spin configurations

$$\mathcal{C}_1 \rightarrow \mathcal{C}_2 \rightarrow
\mathcal{C}_3 \rightarrow \ldots
\rightarrow \mathcal{C}_M$$

where $\mathcal{C}_t = \{ \sigma_1, \ldots, \sigma_N \}$ describes a given
configuration of $N$ spins. Consider these two statements:

1. Two successive configurations are always different $\mathcal{C}_t \ne \mathcal{C}_{t+1}$

2. If $E_t$ is the energy of the spin configuration $\mathcal{C}_t$, then an estimate of the average energy is obtained with $\langle E \rangle \simeq \frac{1}{M} \sum_{t=1}^M E_t$.

[a] Both statements are correct.

[b] Statement (1) is correct and (2) is wrong

[c] Statement (1) is wrong and (2) is correct

[d] Both statements are wrong.

**Question 15** 

Consider the Ising model and in particular the Monte Carlo algorithm based on Wolff cluster moves. 

Which of the following statements is correct?

[a] The Wolff cluster moves do not satisfy the detailed balance, but they do satisfy the more general balance condition.

[b] A configuration will always be changed after a proposed Wolff cluster move.

[c] With Wolff cluster moves, the average cluster size is maximum close to the critical temperature 𝑇𝑐.

[d] During the construction of a cluster, the larger the temperature, the more likely it is that spins are included in the cluster.

**Question 16**

We consider the periodic two-dimensional square lattice Ising model.  The energy for
a given configuration $\{ \sigma \}$ of spins is

$$
E(\{\sigma\}) = -J\sum_{\langle ij \rangle} \sigma_i \sigma_j
$$
where $\langle i j \rangle$ indicates that $i$ and $j$ are
neighbors on the lattice. We use $J=1$ as our unit of energy.
Look at the following programs that intend to compute the total energy and total magnetization of the spin configuration given by `config`:

In [ ]:
import numpy as np
L = 16
config = np.random.choice([-1,1], size=(L,L))
def neighbors(i, j , L):
    return [ ((i+1)%L, j), ((i-1)%L, j), (i, (j+1)%L), (i, (j-1)%L) ]
# Program A
energy = 0.0
for i in range(L):
  for j in range(L):
    for u, v in neighbors(i, j, L):
      energy -= config[i,j] * config[u,v]
magnetization = np.sum(config)
# Program B
energy = 0.0
magnetization = 0.0
for i in range(L):
 for j in range(L):
    for u, v in neighbors(i, j, L):
      energy -= config[i,j] * config[u,v]
      magnetization += config[i,j]
energy /= 2
# Program C
energy = 0.0
magnetization = 0.0
for i in range(L):
  for j in range(L):
    stot = 0.0
    for u, v in neighbors(i, j, L): stot += config[u,v]
    energy -= config[i,j] * stot / 2.0
    magnetization += config[i,j]
# Program D
energy = 0.0
for i in range(L):
  for j in range(L):
    for u, v in neighbors(i, j, L):
      energy = config[i,j] * config[u,v]
energy /= 2
magnetization = np.sum(config)

[a] Program A is correct

[b] Program B is correct

[c] Program C is correct

[d] Program D is correct

**Question  17**

Yous goal is to determine the maximal (average) position reached by random walker that start  at the origin and makes 1000 jumps. At each time step the walker jumps of $+1$ with probability $1/2$ and of $-1$  with probability $1/2$. Look at the program below

In [116]:
import numpy as np
import math
Ntrials = 10000
length =1000
all_max = []
jumps = np.random.choice((1,-1), length )
for sample in range(Ntrials):
    irandom = np.random.randint(length)
    ijump = np.random.choice((1,-1))
    jumps[irandom] = ijump
    walk=np.cumsum(jumps)
    mnew = max(walk)
    if mnew == -1:
       mnew=0
    all_max.append(mnew) 
print("mean maximum=", np.mean(all_max), ",error=", np.std(all_max)/math.sqrt(Ntrials))

mean maximum= 30.6989 ,error= 0.1564320423666456


[a] The mean value and the error are correctly estimated.

[b] The mean value is correctly estimated, but the error is underestimated

[c] The mean value is correctly estimated, but the error is overestimated

[d] The program is not correct as the detailed balance is violated

**Question 18**

Which of the following statements about the *$1/2$ thumb rule* is correct?

[a] It allows to correct the adjustable parameters of a Metropolis algorithm where the detailed balance is weakly violated.

[b] It optimizes the adjustable parameters of a Metropolis algorithm in order to maximize the acceptance rate.

[c] It optimizes the adjustable parameters of a Metropolis algorithm in order to minimize the statistical error.

[d] It is an important rule that ensures the convergence of direct sampling algorithms.

**Question 19**

Consider the simulated annealing algorithm for the Newton spheres problem. During the simulation, the radius of the spheres is gradually increased by some fraction of the maximum allowed radius, namely $R \to R + \gamma (R_\mathrm{max}-R)$. How does the choice for $\gamma$ influence the simulation?

[a] A value of $\gamma$ close to $1$ allows to find a very good optimal packing.

[b] The value of $\gamma$ controls the acceptance rate of proposed moves during the algorithm.

[c] A moderate value for $\gamma$ allows to find a very good optimal packing.

[d] The thumb rule cannot be satisfied if $\gamma$ is taken close to 1.